charms/istio-pilot/src/charm.py

```diff
@@ -251,6 +287,13 @@ def reconcile(self, event):
         # so that we can report them at the end.
         handled_errors = []
 
+        # Call upgrade_charm in case there are new configurations that affect the control plane
+        # This is useful when there is a missing configuration during the install process
+        try:
+            self.upgrade_charm(event)
+        except GenericCharmRuntimeError as err:
+            handled_errors.append(err)
```

Comment:
Is there a reason why this is needed?  I don't have a specific case where I think it breaks, but I worry about it because it combines a few different things.  Telling `istioctl` it is allowed to overwrite istio on every config change, relation change, etc feels like a lot

Reply:
The reason we now have to call upgrade in the reconcile function is because there might be cases where users provide none or only one of the two required cni configurations during install. In that case, we don't want to install the control plane with the cni plugin enabled without these configurations, as we know it will 100% end in error. Instead of us blocking on install and deferring the event in the absence of the required configs, we warn the user that the installation is going to be w/o the plugin and give an option to later provide these configs.

The other option is to defer the install event if these two configs are missing.

Let me know what you think, if you have an alternative, I'll be glad to discuss it.

Reply:
ok, so enable users to take their existing charm and turn on CNI, we need this check in reconcile.  Gotcha  ...  but I'm still nervous :p  

Do we know if `istioctl upgrade` is idempotent (eg: that running it many times will not cause things to restart or interrupt anything)?  I'm searching for docs mentioning this but I don't see anything.  From playing for a few minutes locally I **think** it is idempotent?  I did `istioctl install ...`, did `istioctl upgrade ...` a few times and didn't see anything bad happen.

Reply:
Our [upgrade guide](https://charmed-kubeflow.io/docs/upgrade-17-18#heading--upgrade--istio) tells users to:
* scale down the ingress gateway
* upgrade istio
* scale the ingress back up again
I think that's because of [this note](https://github.com/canonical/istio-operators/blob/main/charms/istio-pilot/README.md#user-content-fn-1-126dacc6f56e23c6bc138b65a6326e5b) about how `istioctl upgrade` sometimes hangs if our ingress gateway exists.  If this is true here, I don't think we can have `upgrade` called in `reconcile`.  But iirc we only saw the hanging upgrade issue when upgrading istio 1.11->1.12 so we might be ok.  I'm not sure how to robustly test it though - I don't recall if it was intermittent or it always occurred.  What do you think?

Reply:
There is a second option, in which we could use the `istioctl install -f <overlay-file.yaml>` to update the configuration the installation. This is how they recommend doing it ([docs](https://istio.io/latest/docs/setup/additional-setup/customize-installation/)).
I'll test that option and update the PR accordingly.

Reply:
I tried a different approach: I am still calling the `upgrade_charm` method in `reconcile`, but the condition to run it is if the configuration that has changed is actually the CNI configurations, that way we avoid running upgrades on every single reconcile. 
The reason why I still want to use `istioctl upgrade` to add these configurations, is because that is the most straightforward way to apply changes while ignoring the things already installed that don't require a change.

What do you think?

Reply:
I like the added check on whether the setting has changed - it adds a tiny bit of complexity but I think its safe and protects against anything bad going on in repeated `istioctl upgrade` calls.  I think you're right that firing a `istioctl upgrade` when things haven't changed is a noop, but nice to just avoid it.

Reply:
I think we can make `_cni_config_changed` generic and put it into chisme then import it here.  Something like (I added StoredState here too just for discussion, but it could be omitted):

```python
def has_config_changed(config_name, storage_method="storedstate"):
  if storage_method == "storedstate":
    current_value = get_it_from_stored_state
  elif storage_method == "peer_relation":
    rel = self.model.get_relation(self.peer_relation_name, None)
    if not rel:
      # If we can't confirm this is a cache hit, this is a cache miss and we return True
      return True
    # Maybe add some prefix on config_name below?
    current = rel.data[self.unit].get(config_name, None)
  new = self.model.config[config_name]
  if new == old:
    # did not change
    return False
  else:
    # did change
    # update cache
    ...
    return True
```

Then in the istio charm, you can do
```python
if has_config_changed("cni-bin-dir") or has_config_changed("cni-conf-dir"):
  do_update()
```

`has_config_changed` can live in chisme and then we can use it elsewhere too

Reply:
I agree this can be a reusable thing, but I prefer to do that refactor in a follow up PR because it will mean the introduction of unrelated changes to this istio cni install effort (things like requirements updates, potentially removal of a unit tests). If you don't mind, I'll do it after we merge this :)

Reply:
I'd have preferred it here (I don't like adding something we immediately plan to remove), but that's ok.  We can resolve this

Reply:
(but lets make an issue, maybe in chisme, saying this could be turned general and maybe linking this thread)